drivers/sensors/pmw3389.c

```diff
@@ -0,0 +1,289 @@
+#define REG_Rest1_Rate_Lower           0x15
+#define REG_Rest1_Rate_Upper           0x16
+#define REG_Rest1_Downshift            0x17
+#define REG_Rest2_Rate_Lower           0x18
+#define REG_Rest2_Rate_Upper           0x19
+#define REG_Rest2_Downshift            0x1A
+#define REG_Rest3_Rate_Lower           0x1B
+#define REG_Rest3_Rate_Upper           0x1C
+#define REG_Observation                0x24
+#define REG_Data_Out_Lower             0x25
+#define REG_Data_Out_Upper             0x26
+#define REG_SROM_ID                    0x2A
+#define REG_Min_SQ_Run                 0x2B
+#define REG_RawData_Threshold          0x2C
+#define REG_Control2                   0x2D
+#define REG_Config5_L                  0x2E
+#define REG_Config5_H                  0x2F
+#define REG_Power_Up_Reset             0X3A
+#define REG_Shutdown                   0x3B
+#define REG_Inverse_Product_ID         0x3F
+#define REG_LiftCutoff_Cal3            0x41
+#define REG_Angle_Snap                 0x42
+#define REG_LiftCutoff_Cal1            0x4A
+#define REG_Motion_Burst               0x50
+#define REG_SROM_Load_Burst            0x62
+#define REG_Lift_Config                0x63
+#define REG_RawData_Burst              0x64
+#define REG_LiftCutoff_Cal2            0x65
+#define REG_LiftCutoff_Cal_Timeout     0x71
+#define REG_LiftCutoff_Cal_Min_Length  0x72
+#define REG_PWM_Period_Cnt             0x73
+#define REG_PWM_Width_Cnt              0x74
+
+// clang-format on
+
+#ifndef MAX_CPI
+#    define MAX_CPI 0x013F  // limits to 0--319, should be max cpi/50 -1
+#endif
+
+bool _inBurst = false;
+
+#ifdef CONSOLE_ENABLE
+void print_byte(uint8_t byte) { dprintf("%c%c%c%c%c%c%c%c|", (byte & 0x80 ? '1' : '0'), (byte & 0x40 ? '1' : '0'), (byte & 0x20 ? '1' : '0'), (byte & 0x10 ? '1' : '0'), (byte & 0x08 ? '1' : '0'), (byte & 0x04 ? '1' : '0'), (byte & 0x02 ? '1' : '0'), (byte & 0x01 ? '1' : '0')); }
+#endif
+#define constrain(amt, low, high) ((amt) < (low) ? (low) : ((amt) > (high) ? (high) : (amt)))
+
+bool spi_start_adv(void) {
+    bool status = spi_start(PMW3389_CS_PIN, PMW3389_SPI_LSBFIRST, PMW3389_SPI_MODE, PMW3389_SPI_DIVISOR);
+    wait_us(1);
+    return status;
+}
+
+void spi_stop_adv(void) {
+    wait_us(1);
+    spi_stop();
+}
+
+spi_status_t spi_write_adv(uint8_t reg_addr, uint8_t data) {
+    if (reg_addr != REG_Motion_Burst) {
+        _inBurst = false;
+    }
+
+    spi_start_adv();
+    // send address of the register, with MSBit = 1 to indicate it's a write
+    spi_status_t status = spi_write(reg_addr | 0x80);
+    status              = spi_write(data);
+
+    // tSCLK-NCS for write operation
+    wait_us(35);
+
+    // tSWW/tSWR (=180us) minus tSCLK-NCS. Could be shortened, but is looks like a safe lower bound
+    wait_us(145);
+    spi_stop();
+    return status;
+}
+
+uint8_t spi_read_adv(uint8_t reg_addr) {
+    spi_start_adv();
+    // send adress of the register, with MSBit = 0 to indicate it's a read
+    spi_write(reg_addr & 0x7f);
+
```

Comment:
I'm half certain that this misses a tRAD sleep here.

Btw, where did you find the datasheet? I only found a 20 page one, not the 57 page one like for the 3360

Reply:
not sure that you need to explicitly wait for tSRAD is this not handled by the spi read? as the read will not finish until the byte is received correct? (which happens after the tSRAD delay)
I am assuming the 20 page document that you are referring to is the same as the one I am using, I just inferred from the changes to the registers and the pmw3360 documentation